logging/src/proc_macros/mod.rs

```diff
@@ -0,0 +1,100 @@
+// This needs to be its own inner crate because of Rust limitations with proc_macros
+// Currently proc_macros cannot export anything other than the macros itself.
+// We need however an export of "slog" because the inner (generated) macros use that
+// That's why we wrap this `proc_macros` crate in the outer `stackable_logging` crate
+// which in turn has a `pub use ::slog;` export.
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::parse::{Parse, ParseStream, Result};
+use syn::{parse_macro_input, ExprLit, Lit, Token};
+
+// Each log statement consists of two things:
+// * A unique code (which must be a number)
+// * The log message
+struct Log {
+    code: ExprLit,
+    msg: ExprLit,
+}
+
+// This parses two literals separated by a comma:
+//     "foo", "bar"
+// While this would parse into `ExprLit` just fine we later verify that the literals
+// are actually of the types we require.
+impl Parse for Log {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let code: ExprLit = input.parse()?;
+        input.parse::<Token![,]>()?;
+        let msg = input.parse()?;
+
+        Ok(Log { code, msg })
+    }
+}
+
+/// This macro can be used to generate a new macro which in turn can be used to log error messages.
+///
+/// ## Usage
+///
+/// ```gen_log_error!(1000, "My templated log message looks {how});```
+///
+/// It accepts two parameters:
+/// * An integer literal as a unique error code
+/// * A string literal (same syntax as the built in `format!` macro
+///
+/// This will generate a new macro which accepts (in this case) two parameters:
+/// ```log_error_1000!(logger, how="fabulous");
+///
+/// The first parameter must be an instance of slog's `Logger` trait, the remaining parameters need to be the template variables.
+/// If they are not provided it will result in a compile-time error.
+#[proc_macro]
+pub fn gen_log_error(input: TokenStream) -> TokenStream {
+    gen_log_macro(input, "error")
+}
+
+// This is the function that takes incoming tokens and generates the new macro
```

Comment:
I'll look at all comment instances again, some I left // on purpose because they are more geared towards developers reading the code and not users using the library.

Reply:
I looked at the generated docs again and none of these appear in there (not pub) so I think this is fine as is.